shell/components/mic-input.js

```diff
@@ -0,0 +1,141 @@
+/*
+@license
+Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+*/
+
+import Xen from './xen/xen.js';
+import Icons from './icons.css.js';
+
+const log = Xen.logFactory('MicInput', 'blue');
+
+class MicInput extends Xen.Base {
+  _getInitialState() {
+    return {
+    };
+  }
+  _didMount() {
+    if (this._supportsSpeech()) {
+      this._initSpeech();
+      document.addEventListener('keydown', this._onKeydown.bind(this));
+      document.addEventListener('keyup', this._onKeyup.bind(this));
+    }
+  }
+  _update(props, state) {
+    if (state.active && !state.recognizing) {
+      this._start();
+    }
+    if (!state.active && state.recognizing) {
+      this._stop();
+    }
+  }
+  _render(props, state) {
+    return state;
+  }
+  _supportsSpeech() {
+    return ('webkitSpeechRecognition' in window);
+  }
+  _initSpeech(state) {
+    const recognition = new window.webkitSpeechRecognition();
+    recognition.continuous = true;
+    recognition.interimResults = true;
+    recognition.onstart = this._onRecognitionStart.bind(this);
+    recognition.onerror = this._onRecognitionError.bind(this);
+    recognition.onend = this._onRecognitionEnd.bind(this);
+    recognition.onresult = this._onRecognitionResult.bind(this);
+    this._setState({recognition});
+  }
+  start() {
+    const {recognition, recognizing, finalTranscript} = this._state;
+    if (recognition) {
+      this._start();
+    } // else complain
```

Comment:
Do we need to complain?

Reply:
Unclear. I left some of these wires exposed until we have more experience with the tech.

Reply:
:+1: